README.md
# InkFilePicker

Ruby API client for filestack.com. Previously known as Ink File Picker and filepicker.io.

[![Build Status](https://travis-ci.org/Skalar/ink_file_picker.svg?branch=master)](https://travis-ci.org/Skalar/ink_file_picker)


## Installation

Add this line to your application's Gemfile:

    gem 'ink_file_picker'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install ink_file_picker



## Usage

This client mirrors part of File Picker's JavaScript API.



### Creating a client

```ruby
# Create a client which will sign URLs. You may drop secret if you have
# not enabled this feature in your developer portal for your application.
client = InkFilePicker.client(key: 'you-api-key', secret: 'your-secret')
```

### Storing a file
```ruby
response = client.store_file file_or_path, content_type
response = client.store_url 'http://www.example.com/img.jpg'
```

### Removing a file
```ruby
response = client.remove url_or_handle_name
```

### Read operations
```ruby
url = client.convert_url url_or_handle_name, w: 100, h: 100
url = client.convert_url url_or_handle_name, {w: 100, h: 100}, expiry: (Time.now + 60 * 10).to_i

# Adds policy and signature, if secret given when client was created.
get_params = {} # Get params we'll be adding to the request.
url = client.retrieve_url url_or_handle_name
url = client.retrieve_url url_or_handle_name, get_params, expiry: (Time.now + 60 * 10).to_i


# Get simple stat on a file, like the Javascript client
stat = client.stat url_or_handle_name

dimentions = client.stat url_or_handle_name, {width: true, height: true}
```

### Errors

When making requests to the API errors may occur. `InkFilePicker::ClientError` or `InkFilePicker::ServerError` will
be raised if we are getting 4xx or 5xx responses back from File Picker. All errors inherits from `InkFilePicker::Error`.

We may also fail with a `InkFilePicker::UnexpectedResponseError`. This happens when for instance you ask File Picker
to download a URL, but the server for the given URL fails to respond within five(?) seconds. We will then get a 200 OK,
but the body will reveal the timeout error in text and the UnexpectedResponseError is raised.
Other download errors may also be in the response, for instance "Invalid response when trying to read from `http://some.url.com/here.jpg`.

## Contributing

1. Fork it (https://github.com/Skalar/ink_file_picker/fork)
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
## 0.0.5
* Added support for Ruby 2.4. (by gudleik)

## 0.0.4
* Dropped ActiveSupport as a dependency. Should make life easier for everyone.
* Expiry on the policy can be given as an object acting as Time, as long as #to_i
  returns Epoch time.

## 0.0.3
* Raise UnexpectedResponseError if we receive a 200 OK, but response body is neither "success", nor valid JSON.
  This may happen if you ask File Picker to download a URL, but the server for the given URL fails to respond
  within five(?) seconds. We will then have a 200 OK, but the body will reveal the timeout error in text and
  the UnexpectedResponseError is raised. Other download errors may also be in the response, for instance
  "Invalid response when trying to read from `http://some.url.com/here.jpg`.

## 0.0.2
* You can call to_hash on response objects.

## 0.0.1
* First release. Main functionality like store_url, store_file, remove, stat etc.


lib/ink_file_picker/client.rb
require "faraday"

module InkFilePicker
  class Client
    attr_accessor :configuration

    def initialize(configuration)
      self.configuration = Configuration.new configuration
    end


    # Public: Store a file from given URL.
    #
    # url                 - URL to resource
    # policy_attributes   - If you use security policies you may send in for instance {e: (Time.now + 60 * 10)} here
    #
    # Returns a hash representing the response where you can read for instance 'url'
    def store_url(url, policy_attributes = {})
      params = {key: configuration.key}

      add_policy_to params, from: policy_attributes, ensure_included: {call: 'store'}

      response = http_connection.post configuration.store_path do |request|
        request.params = params
        request.body = {url: url}
      end

      wrap_response_or_fail_unless_success! response
    end

    # Public: Store a file from given local file or path.
    #
    # file_or_path        - File or path to file
    # content_type        - The file's content type
    # filename            - The file's name, optional
    # policy_attributes   - If you use security policies you may send in for instance {expire: (Time.now + 60 * 10)} here
    #
    # Returns a hash representing the response where you can read for instance 'url'
    def store_file(file_or_path, content_type, filename = nil, policy_attributes = {})
      file_upload = Faraday::UploadIO.new file_or_path, content_type, filename
      params = {key: configuration.key}

      add_policy_to params, from: policy_attributes, ensure_included: {call: 'store'}

      response = http_connection.post configuration.store_path do |request|
        request.params = params
        request.body = {fileUpload: file_upload}
      end

      wrap_response_or_fail_unless_success! response
    end

    # Public: Removes a file from file picker.
    #
    # handle_or_url       - The handle or URL to the file
    # policy_attributes   - If you use security policies you may send in for instance {expire: (Time.now + 60 * 10)} here
    #
    # Returns boolean value
    def remove(handle_or_url, policy_attributes = {})
      response = http_connection.delete remove_url(handle_or_url, policy_attributes)

      wrap_response_or_fail_unless_success! response
    end


    # Public: Returns short stat for a file
    #
    # handle_or_url       - The handle or URL to the file
    # params              - Request params, like {width: true, height: true} to get width and height info. May be empty for default response
    # policy_attributes   - If you use security policies you may send in for instance {expire: (Time.now + 60 * 10)} here
    #
    # Returns hash of headers returned from file picker or false if request was unsuccessful
    def stat(handle_or_url, params = {}, policy_attributes = {})
      response = http_connection.get stat_url(handle_or_url, params, policy_attributes)

      wrap_response_or_fail_unless_success! response
    end

    # Public: Generates a you can use for removing an asset on file picker.
    #
    # handle_or_url       - The handle or URL to the file
    # policy_attributes   - If you use security policies you may send in for instance {expire: (Time.now + 60 * 10)} here
    #
    # Returns a URL to the converted image
    def remove_url(handle_or_url, policy_attributes = {})
      generate_url handle_or_url, {key: configuration.key}, policy_attributes, call: 'remove'
    end

    # Public: Generates a convert URL for given file.
    #
    # handle_or_url       - The handle or URL to the file
    # params              - Convert params, like {w: 100, h:100}
    # policy_attributes   - If you use security policies you may send in for instance {expire: (Time.now + 60 * 10)} here
    #
    # Returns a URL to the converted image
    def convert_url(handle_or_url, params = {}, policy_attributes = {})
      generate_url handle_or_url, params, policy_attributes, call: 'convert', url_action: 'convert'
    end


    # Public: Generates a URL for a given file
    #
    # handle_or_url       - The handle or URL to the file
    # params              - Params to be added as get params, like {cache: true}
    # policy_attributes   - If you use security policies you may send in for instance {expire: (Time.now + 60 * 10)} here
    #
    # This method is not that usefull unless you have enabled security policy
    #
    # Returns a URL to the image
    def retrieve_url(handle_or_url, params = {}, policy_attributes = {})
      generate_url handle_or_url, params, policy_attributes, call: 'read'
    end


    # Public: Generates a stat URL for a given file
    #
    # handle_or_url       - The handle or URL to the file
    # policy_attributes   - If you use security policies you may send in for instance {expire: (Time.now + 60 * 10)} here
    #
    # Returns a URL to the image you can do a HEAD request to in order to get stats
    def stat_url(handle_or_url, params, policy_attributes = {})
      generate_url handle_or_url, params, policy_attributes, call: 'stat', url_action: 'metadata'
    end





    # Public: Creates a policy with default configuration set in this client.
    #
    # Returns Policy object
    def policy(attributes)
      defaults = {
        secret: configuration.secret,
        expiry: Time.now.to_i + configuration.default_expiry
      }.freeze

      attributes = defaults.merge attributes

      Policy.new attributes
    end





    def http_connection
      @http_connection ||= Faraday.new(url: configuration.filepicker_url) do |builder|
        builder.request :multipart
        builder.request :url_encoded
        builder.adapter configuration.http_adapter || Faraday.default_adapter
      end
    end


    private

    def generate_url(handle_or_url, params, policy_attributes, options)
      file_handle = FileHandle.new handle_or_url, configuration.cdn_url

      add_policy_to params, from: policy_attributes, ensure_included: {handle: file_handle.handle, call: options[:call]}

      UrlBuilder.new(file_url: file_handle.url, action: options[:url_action], params: params).to_s
    end

    def add_policy_to(params, options = {})
      policy_attributes = (options[:from] || {}).merge options[:ensure_included]
      params.merge! policy(policy_attributes)
    end

    # Private: Inspects response for error and raise a InkFilePicker error if client/server error.
    def wrap_response_or_fail_unless_success!(response)
      case response.status
      when 200...300
        Response.new(response).tap do |wrapped_response|
          fail UnexpectedResponseError, response.body unless wrapped_response.valid?
        end
      when 400...500
        fail ClientError.new response.body, response
      when 500...600
        fail ServerError.new response.body, response
      else
        fail Error, "Response was neither a success, nor within http status 400...600. Response was: '#{response.inspect}'."
      end
    end
  end
end
